FirebaseMLModelDownloader/Sources/ModelDownloader.swift

```diff
@@ -283,25 +286,81 @@ extension ModelDownloader {
       appName: appName,
       localModelInfo: localModelInfo
     )
+    let downloader = ModelFileDownloader(conditions: conditions)
+    downloadInfoAndModel(
+      modelName: modelName,
+      modelInfoRetriever: modelInfoRetriever,
+      downloader: downloader,
+      conditions: conditions,
+      progressHandler: progressHandler,
+      completion: completion
+    )
+  }
+
+  func downloadInfoAndModel(modelName: String,
+                            modelInfoRetriever: ModelInfoRetriever,
+                            downloader: FileDownloader,
+                            conditions: ModelDownloadConditions,
+                            progressHandler: ((Float) -> Void)? = nil,
+                            completion: @escaping (Result<CustomModel, DownloadError>)
+                              -> Void) {
     modelInfoRetriever.downloadModelInfo { result in
       switch result {
       case let .success(downloadModelInfoResult):
         switch downloadModelInfoResult {
         /// New model info was downloaded from server.
         case let .modelInfo(remoteModelInfo):
-          let downloader = ModelFileDownloader(conditions: conditions)
           let downloadTask = ModelDownloadTask(
             remoteModelInfo: remoteModelInfo,
             appName: self.appName,
             defaults: self.userDefaults,
             downloader: downloader,
-            modelInfoRetriever: modelInfoRetriever,
-            telemetryLogger: self.telemetryLogger,
-            progressHandler: progressHandler,
-            completion: completion
+            telemetryLogger: self.telemetryLogger
           )
           // TODO: Is it possible for download URL to expire here?
-          downloadTask.download()
+          downloadTask.download(progressHandler: { progress in
+            DispatchQueue.main.async {
+              progressHandler?(progress)
+            }
+          }) { result in
+            switch result {
+            case let .success(model):
+              DispatchQueue.main.async {
+                completion(.success(model))
+              }
+            case let .failure(error):
+              switch error {
+              case .failedPrecondition:
```

Comment:
A couple clarification comments on the checks below may help in the future. Alternatively we may refactor the checks to a method like `shouldRetryDownload(urlExpiryTime:, numberOfRetries:)` to make the code a bit more self-documenting.

Reply:
I didn't do the refactor, but I added some comments. Can look at refactoring in a future PR.